cardano-node/src/Cardano/Node/Configuration/POM.hs

```diff
@@ -0,0 +1,328 @@
+         -- all the configuration settings required for the cardano node
+         -- (logging, tracing, protocol, slot length etc)
+       , pncConfigFile      :: !(Last ConfigYamlFilePath)
+       , pncTopologyFile    :: !(Last TopologyFile)
+       , pncDatabaseFile    :: !(Last DbFile)
+       , pncProtocolFiles   :: !(Last ProtocolFilepaths)
+       , pncValidateDB      :: !(Last Bool)
+       , pncShutdownIPC     :: !(Last (Maybe Fd))
+       , pncShutdownOnSlotSynced :: !(Last MaxSlotNo)
+
+          -- Protocol-specific parameters:
+       , pncProtocolConfig :: !(Last NodeProtocolConfiguration)
+
+         -- Node parameters, not protocol-specific:
+       , pncSocketPath     :: !(Last SocketPath)
+
+         -- BlockFetch configuration
+       , pncMaxConcurrencyBulkSync :: !(Last MaxConcurrencyBulkSync)
+       , pncMaxConcurrencyDeadline :: !(Last MaxConcurrencyDeadline)
+
+         -- Logging parameters:
+       , pncViewMode       :: !(Last ViewMode)
+       , pncLoggingSwitch  :: !(Last Bool)
+       , pncLogMetrics     :: !(Last Bool)
+       , pncTraceConfig    :: !(Last TraceOptions)
+       } deriving (Eq, Generic, Show)
+
+instance AdjustFilePaths PartialNodeConfiguration where
+  adjustFilePaths f x@PartialNodeConfiguration {
+                        pncProtocolConfig,
+                        pncSocketPath
+                      } =
+    x {
+      pncProtocolConfig = adjustFilePaths f pncProtocolConfig,
+      pncSocketPath     = adjustFilePaths f pncSocketPath
+    }
+
+instance Semigroup PartialNodeConfiguration where
+  (<>) = gmappend
+
+instance Monoid PartialNodeConfiguration where
+  mempty = PartialNodeConfiguration
+             mempty mempty mempty mempty mempty
+             mempty mempty mempty mempty mempty
+             mempty mempty mempty mempty mempty mempty
+  mappend x y =
+    PartialNodeConfiguration
+      { pncNodeAddr = pncNodeAddr x <> pncNodeAddr y
+      , pncConfigFile = pncConfigFile x <> pncConfigFile y
+      , pncTopologyFile = pncTopologyFile x <> pncTopologyFile y
+      , pncDatabaseFile = pncDatabaseFile x <> pncDatabaseFile y
+      , pncProtocolFiles = pncProtocolFiles x <> pncProtocolFiles y
+      , pncValidateDB = pncValidateDB x <> pncValidateDB y
+      , pncShutdownIPC = pncShutdownIPC x <> pncShutdownIPC y
+      , pncShutdownOnSlotSynced = pncShutdownOnSlotSynced x <> pncShutdownOnSlotSynced y
+      , pncProtocolConfig = pncProtocolConfig x <> pncProtocolConfig y
+      , pncSocketPath = pncSocketPath x <> pncSocketPath y
+      , pncMaxConcurrencyBulkSync = pncMaxConcurrencyBulkSync x <> pncMaxConcurrencyBulkSync y
+      , pncMaxConcurrencyDeadline = pncMaxConcurrencyDeadline x <> pncMaxConcurrencyDeadline y
+      , pncViewMode = pncViewMode x <> pncViewMode y
+      , pncLoggingSwitch = pncLoggingSwitch x <> pncLoggingSwitch y
+      , pncLogMetrics = pncLogMetrics x <> pncLogMetrics y
+      , pncTraceConfig = pncTraceConfig x <> pncTraceConfig y
+      }
```

Comment:
Shouldn't `(<>)` and `mappend` do the same thing?

Reply:
Yep, I implemented the `Monoid` instance to use the notation `mempty { ... }` to avoid writing `mempty` for each record didn't have a value. I've removed the instance because being explicit makes it clearer what I'm doing.